🚀 2단계 - 예외 처리 및 리팩터링

hhhhhwi opened the pull request:
안녕하세요! 2단계 PR을 이제서야 😢 올립니다.

예외 처리를 하는 과정에서 생각보다 수정이 많고 시간이 오래걸렸습니다 ㅠ
비즈니스 로직을 수정하는 데에도 어느정도 시간이 걸렸지만,
어떤 예외를, 어떻게 응답을 하는 것이 최선일 지 고민하는 데에 시간이 가장 많이 걸렸던 것 같습니다.

비즈니스 로직에 대한 예외는 대부분 Custom Exception을 만들어서 호출하는 방향으로 구현했는데, 구현하다보니 각 예외마다 Exception을 만들고 있고 유지보수 관점에서 너무 많은 Custom Exception이 관리하기 어렵게 만들지 않을까? 라는 생각이 들었습니다.

그리고 Exception 호출 시 응답객체에 대한 타입도 고민이 되었는데요..! HttpStatus 코드 값을 전달해주는데 응답 값으로 한 번 더 전달해줄 필요가 있을까? 라는 생각에 우선 리턴 메시지와 리턴 메시지에 정의된 enum 값을 전달해주는 방향으로 구현해보았습니다..!

1단계 PR에서 제안해주신 피드백도 적용해보았는데 제가 맞게 이해를 하고 적용했는지 모르겠네요 ㅎㅎ

이번 피드백도 전달해주시면, 잘 반영해보겠습니다!

감사합니다😄

boorownie committed:
fix: theme 추가로 인한 링크 추가 및 기존 링크 수정

(cherry picked from commit fe7a701d48e0dd25d7d2502a3deee442db944c2b)

boorownie committed:
feat: theme 추가로 인한 html 및 js 파일 생성

(cherry picked from commit 97663d962984ffd9f557bf82ef9767926d0dea84)

hhhhhwi committed:
refactor: 테스트 분리 및 가독성을 위한 메소드 명 수정

hhhhhwi committed:
refactor: properties -> yml 파일로 수정

hhhhhwi committed:
refactor: 가독성을 위해 stringBuilder로 sql 작성

hhhhhwi committed:
refactor: RequestMapping 적용

hhhhhwi committed:
docs: 테마 관련 api 명세 추가

hhhhhwi committed:
refactor: Domain <-> DTO 간 변환 작업의 책임 변경

hhhhhwi committed:
feat: 테마 관리 기능 추가

hhhhhwi committed:
refactor: 예약 등록 및 조회 시, 테마 관련 내용 추가

hhhhhwi committed:
feat: 전역 예외 처리 적용

hhhhhwi committed:
refactor: validator를 사용하여 입력 값에 대한 검증 및 exception 처리

hhhhhwi committed:
refactor: Controller 어노테이션을 RestController 어노테이션으로 변경함으로서 Json 형태의 객체로 자동 변환

hhhhhwi committed:
refactor: save시, 등록한 객체를 반환하도록 수정

hhhhhwi committed:
feat: 이전 날짜에 대한 예약 등록 시, 예외 호출하는 로직 추가
- Custom Exception 구현 : PastDateTimeExeption

hhhhhwi committed:
refactor: 실패 테스트 케이스 추가 및 화면 수정

hhhhhwi committed:
refactor: 예약하려는 날짜, 시간에 해당 테마가 이미 예약되어 있는 경우 예외 호출
- custom exception : AlreadyExistsException

hhhhhwi committed:
refactor: 삭제하려는 시간, 테마에 대한 예약이 있을 경우 예외 호출
- custom exception : ReferentialIntegrityException

pkeugine commented on src/main/java/roomescape/exception/NotExistsException.java:
```diff
@@ -0,0 +1,18 @@
+package roomescape.exception;
+
+public class NotExistsException extends RuntimeException{
```
```suggestion
public class NotExistsException extends RuntimeException {
```

소소한 디테일이지만, [이전 단계에서 얘기했던 컨벤션 내용](https://github.com/next-step/spring-roomescape-reservation/pull/56#discussion_r1629354186)이라 한 번 더 코멘트 달아봅니다!
intelliJ 기능 중 코드를 원하는 컨벤션에 맞게 수정해주는 기능이 있는데, 한 번 알아보고 사용해봐도 좋을 것 같아요.
이런건 누구나 못 보고 넘어갈 수 있으니까요 ㅋㅋㅋ

(+) 아직 space indent 가 아닌 tab indent 네요!

hhhhhwi replied:
제가 깜빡하고 적용을 못 했었네요 ㅎㅎ 
유근님이 추천해주신 spring-javaformat 플러그인도 적용해보고 싶었는데
우선 https://github.com/google/styleguide 활용해서 구글 컨벤션 적용해보았습니다!

pkeugine commented on src/main/java/roomescape/exception/NotExistsException.java:
```diff
@@ -0,0 +1,18 @@
+
+	public NotExistsException() {
+		this("");
+	}
```
사용하지 않는 로직은 제거해주는게 어떨까요?

pkeugine commented on src/main/java/roomescape/exception/RoomescapeErrorCode.java:
```diff
@@ -0,0 +1,20 @@
+package roomescape.exception;
+
+public enum RoomescapeErrorCode {
```
해당 enum 이름이 적절한지 생각해봐도 좋을 것 같아요.
error code 를 관리하는 부분인가요?

hhhhhwi replied:
ExceptionResponse에 해당 Enum을 반환하면서 해당 Enum 값이 코드로서 활요할 수 있다고 판단하여 해당 클래스 명을 붙였는데요..! 사실 필요 없는 응답값이기도 하고, 에러메시지 정의가 더 목적에 맞는 부분인 것 같습니다 😄

pkeugine commented on src/main/java/roomescape/exception/RoomescapeErrorCode.java:
```diff
@@ -0,0 +1,20 @@
+package roomescape.exception;
+
+public enum RoomescapeErrorCode {
+	ILLEGAL_INPUT_VALUE_EXCEPTION (" 값을 확인해주세요."),
+	ILLEGAL_DATETIME_FORMAT_EXCEPTION ("날짜 및 시간 입력 형식을 확인해주세요."),
+	PAST_DATETIME_EXCEPTION("과거 날짜 및 시간을 예약할 수 없습니다."),
+	EXISTS_EXCEPTION(" 이미 존재합니다."),
+	NOT_EXISTS_EXCEPTION(" 존재하지 않습니다."),
+	REFERENTIAL_EXCEPTION("이/가 존재하여 데이터를 수정할 수 없습니다.");
```
에러 메시지를 한 곳에서 관리하도록 하셨군요!
이 방식의 장점과 단점은 무엇일까요?

hhhhhwi replied:
에러 메시지의 관리 포인트가 한 곳에서 이루어지기 때문에 유지보수가 용이하다는 장점이 있다고 생각합니다..!
단점으로는... 동적으로 변경하기 어렵다..? 정도만 생각나는데요..! 

에러메시지에 대한 유진님의 의견이 궁금하네요..! 주로 어떤 방식으로 관리하시는 편이신가요?!

pkeugine replied:
사실 이건 정답이 있는 문제는 아니라서 장단점을 물어보는 식으로 코멘트를 남기긴 했는데요,
저는 모으지 않고 그냥 두거나, 커스텀 예외 안에서 관리하는 편이에요.
이게 생각과는 다르게 한 곳에 예외 관련 관리포인트가 모여지지 않더라구요.
중복으로 생기는 것들도 많고, 내가 사용하고자 하는 것을 찾기 힘들 때도 있어요. 예외가 나중에는 엄청 많아지니까요.
그래도 이 방식을 한 번 사용해보며, 최대한 그런 일이 발생하지 않도록 하는 것도 좋은 시도라고 생각합니다.
앞서 말했듯이 틀린 방식은 아니니까요!

pkeugine commented on src/main/java/roomescape/exception/RoomescapeExceptionHandler.java:
```diff
@@ -0,0 +1,55 @@
+	@ExceptionHandler(MethodArgumentNotValidException.class)
+	@ResponseStatus(HttpStatus.BAD_REQUEST)
+	public RoomescapeExceptionResponse handleMethodArgumentNotValidException(MethodArgumentNotValidException e) {
+		return RoomescapeExceptionResponse.from(RoomescapeErrorCode.ILLEGAL_INPUT_VALUE_EXCEPTION, e.getBindingResult().getFieldError().getDefaultMessage());
```
휘연님은 가로 제한을 몇으로 두고 개발하시나요?
코드가 너무 길어져서 intelliJ 에서 보이는 우측 회색 선을 넘긴다면 줄바꿈을 활용해서 가독성을 높여볼 수 있을 것 같아요.
<img width="1234" alt="Screenshot 2024-06-12 at 3 45 33 AM" src="https://github.com/next-step/spring-roomescape-reservation/assets/48251668/0226792e-9301-4409-b069-105f9e220399">

hhhhhwi replied:
ㅎㅎ 제가 이번에 너무 코드 작성을 형편없이 했네요;; 정신차리고 보니 깜짝 놀랬습니다 😓

pkeugine replied:
형편없다고 생각하진 않았습니다 ㅋㅋㅋ
저는 이왕이면 옆으로 긴 것보단 위아래로 긴게 훨씬 보기 좋더라구요.

pkeugine commented on src/main/java/roomescape/exception/PastDateTimeExeption.java:
```diff
@@ -0,0 +1,18 @@
+package roomescape.exception;
+
+public class PastDateTimeExeption extends RuntimeException{
```
```suggestion
public class PastDateTimeException extends RuntimeException {
```

pkeugine commented on src/main/java/roomescape/reservation/ReservationService.java:
```diff
@@ -1,30 +1,57 @@
-		return reservationRepository.save(request.getName(), request.getDate(), request.getTimeId());
+	public ReservationResponse saveReservation(ReservationRequest request) {
+		ReservationTime reservationTime = Optional.ofNullable(reservationTimeRepository.findById(request.getTimeId()))
+				.orElseThrow(() -> new NotExistsException("해당 시간이"));
+		Theme theme = Optional.ofNullable(themeRepository.findById(request.getThemeId()))
+				.orElseThrow(() -> new NotExistsException("해당 테마가"));
```
시간과 테마가 없을 때 각각 `NotExistsException` 을 사용하는군요!
좋은 시작점인 것 같아요.
각 예외를 사용할 때마다 `해당 시간이` 와 같은 메시지를 매번 넣어주기보단,
`NotExistsException` 을 상속받는 `ReservationTimeNotExistsException` 과 `ThemeNotExistsException` 을 만들어보는건 어떨까요? 이 방식을 사용해서 직접 예외 메시지의 일부를 넣어주지 않아도 원하는 내용이 출력되게 할 수 있을 것 같아요.

hhhhhwi replied:
오..! CustomException을 구현했음에도 불구하고 각자 다른 메시지를 출력하는 부분이 마음에 걸렸었는데, 상속을 사용하면 더 구체적인 예외를 만들 수 있겠네요!

pkeugine commented on src/main/java/roomescape/exception/ReferentialIntegrityException.java:
```diff
@@ -0,0 +1,18 @@
+package roomescape.exception;
+
+public class ReferentialIntegrityException extends RuntimeException{
```
> 구현하다보니 각 예외마다 Exception을 만들고 있고 유지보수 관점에서 너무 많은 Custom Exception이 관리하기 어렵게 만들지 않을까? 라는 생각이 들었습니다.

예외가 많아질수록 중복되는 상황을 나타내는 예외가 생길 수도 있고, 그렇다고 내가 처한 상황을 나타내는 예외가 있는지 하나씩 확인해보는 것도 오래 걸리죠. 이런 상황을 최대한 피하려고 예외 클래스 이름을 잘 짓는다던지, 예외간 계층 구조를 만들어본다던지, 아니면 패키지 정리를 하여 그나마 보기 쉽게 하려해요. 그럼에도 그 종류가 많아지는건 커스텀 예외를 사용한다면 언젠간 발생하는 요소인 것 같네요.

그런데 저는 막상 프로젝트를 유지보수하다보면 커스텀 예외가 많아서 느끼는 불편함이 그리 크진 않은 것 같아요. 그런데 이건 사람마다 다를거라.. 이 기회에 한 번 커스텀 예외 전략을 세워보면서 직접 느껴보는 것도 재밌겠네요 :)

pkeugine commented on src/main/java/roomescape/exception/RoomescapeExceptionHandler.java:
```diff
@@ -0,0 +1,55 @@
+
+@RestControllerAdvice
+public class RoomescapeExceptionHandler {
+	@ExceptionHandler(IllegalArgumentException.class)
+	@ResponseStatus(HttpStatus.BAD_REQUEST)
+	public RoomescapeExceptionResponse IllegalArgumentException(IllegalArgumentException e) {
+		return RoomescapeExceptionResponse.from(RoomescapeErrorCode.ILLEGAL_INPUT_VALUE_EXCEPTION, e.getMessage());
+	}
```
> HttpStatus 코드 값을 전달해주는데 응답 값으로 한 번 더 전달해줄 필요가 있을까? 라는 생각에 우선 리턴 메시지와 리턴 메시지에 정의된 enum 값을 전달해주는 방향으로 구현해보았습니다..!

좋은 판단이에요!
사실 내려주는 정보는 client 와 server 관리자간 협의를 통해 만들어지는거라, '그럼에도 불구하고 status code 도 필요하다' 고 말하면 주는거죠... ㅎ 지금은 그런 상황은 아니니, 말해주신 이유 때문에 빼는게 전혀 문제되지 않는다고 생각해요. 저는 더 나아가 enum 정보도 없이 message 만 보내도 괜찮다고 봐요! api 의 정상적인 응답처럼 예외도 형식만 갖춰져 있다면 괜찮습니다!

pkeugine commented on src/main/java/roomescape/exception/RoomescapeExceptionHandler.java:
```diff
@@ -0,0 +1,55 @@
+		return RoomescapeExceptionResponse.from(RoomescapeErrorCode.ILLEGAL_INPUT_VALUE_EXCEPTION, e.getBindingResult().getFieldError().getDefaultMessage());
+	}
+
+	@ExceptionHandler(DuplicateRequestException.class)
+	@ResponseStatus(HttpStatus.CONFLICT)
+	public RoomescapeExceptionResponse handleDuplicateRequestException(DuplicateRequestException e) {
+		return RoomescapeExceptionResponse.from(RoomescapeErrorCode.EXISTS_EXCEPTION, e.getMessage());
+	}
```
`@ResponseStatus` 가 아니라 이 곳도 controller 처럼 `ResponseEntity` 를 사용해보면 어떨까요?

hhhhhwi replied:
혹시 `@ResponseStatus`대신 `ResponseEntity` 사용을 권해주신 이유가 무엇일까요?!

`@RestControllerAdvice` 는 `@ControllerAdvice `어노테이션과 `@ResponseBody` 어노테이션이 합쳐진 것이라고 알고 있는데요..! 응답값이 body로 세팅되어 `ResponseEntity`를 사용하지 않아도 판단해서 위와 같이 구현했는데 혹시 다른 이유가 있을까 싶어서요!

pkeugine replied:
네 필수는 아닙니다. `ResponseEntity` 를 사용하면 정상적인 응답과 에러가 발생했을 때의 응답 로직을 통일할 수도 있고, header 같은 것들을 추가할 수 있는 유연함이 있어서 저는 선호해요. 익숙하기 때문에 읽기도 좋고, 유연하기까지 하다면 꽤나 좋은 선택지라고 보는거죠!

pkeugine commented on src/main/java/roomescape/reservation/ReservationController.java:
```diff
@@ -1,34 +1,32 @@
 
-@Controller
+@RestController
+@RequestMapping("/reservations")
```
리뷰 잘 반영해주셨네요 👍

pkeugine commented on src/test/java/roomescape/ReservationAcceptanceTest.java:
```diff
@@ -0,0 +1,129 @@
+		ExtractableResponse<Response> response = 예약_등록(request);
+
+		assertThat(response.statusCode()).isEqualTo(200);
+	}
```
`@DisplayName` 을 잘 사용해주셨네요!! 👍 
해당 어노테이션은 메서드 단위로도 적용할 수 있어서, 테스트 메서드명을 영어로 쓰고 한글 이름을 붙일 수 있어요.
이건 휘연님이 더 편하신 방법으로 해도 괜찮을 것 같습니다.

pkeugine commented on src/test/java/roomescape/ReservationAcceptanceTest.java:
```diff
@@ -0,0 +1,129 @@
+	void 예약_등록_성공() {
+		ExtractableResponse<Response> response = 예약_등록(request);
+
+		assertThat(response.statusCode()).isEqualTo(200);
```
저는 이왕이면 테스트가 어떻게 작동되는지 '테스트 메서드' 안에서 보면 좋을 것 같은데, 휘연님은 어떻게 생각하시나요?

hhhhhwi replied:
저는 인수테스트 코드는 개발자들이 요구사항에 대한 파악을 빠르게 하기 위한 목적이 크다고 생각하여
최대한 공통 메서드를 사용하여 한 눈에 파악할 수 있도록 구현하는 편입니다..!
메서드 명을 한글로 하는 이유 또한 위와 같습니다!

유진님 말씀처럼 지금처럼 짧은 api 호출 정도는 하나의 메서드 안에서 읽는 데에 어려움이 없지만,
조건이 많아지고, 테스트 픽스쳐가 늘어날 수록 테스트 내용을 파악하기 어려워질 수 있다고 생각합니다!

pkeugine replied:
맞습니다! 특히 테스트 픽스쳐가 늘어나는 경우가 거의 대부분이라서 골치가 아프기도 해요.
그런데 테스트가 많아졌을 때, 테스트끼리 공유하는 메서드, 픽스쳐, 또는 변수가 많아질수록 테스트를 변경하기 어려워진다는 점 또한 염두에 둘 필요는 있어요. 테스트는 많이들 프로덕션 로직과는 별개인 '부가적인' 것으로 생각하는 경우가 많아서, 그렇게 변경하기 복잡하다면 점점 건드리지 않는... 특징이 있더라구요. 물론 그 상황을 애초에 만들지 않고 잘 만들면 되긴 합니다.

그래도 좋은 답변이었어요!! 👍
이렇게 두고, 한 번 잘 관리해보도록 하죠!

pkeugine commented on src/test/java/roomescape/ReservationAcceptanceTest.java:
```diff
@@ -0,0 +1,129 @@
+		assertThat(response.jsonPath().getList("name")).contains(name);
+	}
+
+	@Test
+	void 예약_삭제_성공() {
+		예약_등록(request);
+
+		ExtractableResponse<Response> responseBeforeDelete = 예약_조회();
+
+		assertThat(responseBeforeDelete.statusCode()).isEqualTo(200);
+		assertThat(responseBeforeDelete.jsonPath().getList("name")).hasSize(1);
+
+		RestAssured.given().log().all()
+				.when().delete("/reservations/1")
+				.then().log().all()
+				.statusCode(200);
+
+		ExtractableResponse<Response> responseAfterDelete = 예약_조회();
+
+		assertThat(responseAfterDelete.statusCode()).isEqualTo(200);
+		assertThat(responseAfterDelete.jsonPath().getList("name")).hasSize(0);
+	}
```
예약 삭제를 확인하는 테스트에서 '등록 후 조회' 를 검증할 필요가 있을까요?
그냥 등록을 한 뒤, 그것을 지우고 내려오는 response 를 검증만해도 충분할 것 같아서요!
마지막에 조회를 하여 값이 없는 것을 확인하는 것도 저는 불필요하다고 봐요.
검증하고자 하는 것만 확실하게 검증하고 그 외의 것들은 테스트에서 빼보도록하죠!

hhhhhwi replied:
전과 후를 비교하는 부분이 필요하다고 생각했는데, 검증하고자 하는 지점만 확인하는 게 더 깔끔할 것 같네요! 👍

pkeugine commented on src/test/java/roomescape/ThemeAcceptanceTest.java:
```diff
@@ -0,0 +1,79 @@
+@DisplayName("테마 관련 api 호출 테스트")
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT)
+@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
+public class ThemeAcceptanceTest {
```
새로운 기능에 대한 테스트 좋네요 👍 
그리고 각 상황을 잊지 않고 다 테스트해주셨어요! 💯

pkeugine commented on src/main/java/roomescape/theme/ThemeController.java:
```diff
@@ -0,0 +1,36 @@
+
+	@PostMapping
+	public ResponseEntity<ThemeResponse> saveThemes(@RequestBody @Valid ThemeRequest request) {
+		Long id = themeService.saveThemes(request);
```
복수형은 여러 theme 을 저장하는 경우에 사용해보면 좋을 것 같아요!
예를 들어 위에 있는 `getThemes` 는 list 형태로 여러 개를 내려주고 있어요.

pkeugine commented on src/main/java/roomescape/theme/ThemeController.java:
```diff
@@ -0,0 +1,36 @@
+
+	@GetMapping
+	public List<ThemeResponse> getThemes() {
+		return themeService.findThemes();
```
응답의 http status code 나 header 를 수정하지 않더라도 모든 api 에 대해 `ResponseEntity` 를 사용해서 응답을 만들어보는건 어떨까요?

hhhhhwi replied:
위의 질문과 마찬가지로 @ResponseBody 가 포함된 @RestController 를 사용하는 경우에도 ResponseEntity 사용을 추천하시는 이유가 있으실까요?

hhhhhwi replied:
하나의 컨트롤러에서 응답 유형을 하나로 맞추기 위함일까요?

그렇다면 RestController 어노테이션을 사용하는 또 다른 이유가 있을까요?

혼자서 고민해본 결과로는,
`@RestController` 의 역할이 `@ResponseBody`에 대한 역할보다는, Api를 지원하는 Controller 임을 표현하는 의미가 있다고 생각이 들었습니다..!

코드 일관성 또한 필요하다고 생각이 들어, 해당 프로젝트의 RestController 반환 유형을 ResponseEntity로 수정해보았습니다!

pkeugine replied:
제가 `ResponseEntity` 를 꽤 좋아하는 편인 것 같아요 ㅋㅋㅋ
유연하게 응답을 만들 수 있고, 읽기도 편하죠. 굉장히 잘 만든 api 라고 생각합니다.
답변해주신 것처럼 `@RestController` 라는 어노테이션이 붙으면 저 역시 대부분 애플리케이션의 api 를 지원하는 것으로 받아들이네요. 좋은 포인트입니다. 🤔

pkeugine reviewed (changes requested):
안녕하세요 휘연님!
이번에도 잘 진행해주셨네요.
몇 가지 리뷰 남겨보았으니 한 번 확인해주세요!!
저는 정말 괜찮으니 리뷰 반영하시면서 생기는 질문이나 이해가 잘 안되는 리뷰가 있다면
언제든지 DM 주시면 됩니다. :)

hhhhhwi committed:
refactor: 코드 컨벤션 적용 및 가독성 개선을 위한 줄바꿈

hhhhhwi committed:
refactor: 불필요한 검증 테스트 코드 삭제

hhhhhwi committed:
refactor: NotExistsException 상속으로 ReservationTimeNotExistsException, ThemeNotExistsException 구현. 예외 구체화.

hhhhhwi committed:
refactor: 클래스명 변경 및 필요없는 응답 값 삭제
- RoomescapeErrorCode -> RoomescapeErrorMessage
- RoomescapeExceptionResponse의 code 삭제

hhhhhwi committed:
refactor: 패키지 변경

hhhhhwi committed:
refactor: ResponseEntity를 통해 응답하도록 수정

hhhhhwi commented:
안녕하세요 유진님!
정성스럽게 작성해주신 피드백 감사합니다!
덕분에 놓치고 있던 부분이 많이 보였습니다 👍 

반영하면서 궁금했던 부분은 코멘트로 추가해보았습니다..!

혹시 제가 놓친 피드백이나 추가적인 피드백이 있으시면 바로 말씀해주세요 😓 
챙긴다고 챙겼는데.. 계속 놓치는 부분이 생기더라구요 ㅠㅠ

늘 감사합니다!

pkeugine commented on src/main/java/roomescape/reservation/Reservation.java:
```diff
@@ -1,24 +1,57 @@
 
-	private ReservationTime reservationTime;
+    public ReservationTime getReservationTime() {
+        return reservationTime;
+    }
 
-	public Reservation(Long id, String name, String date, ReservationTime reservationTime) {
-		this.id = id;
-		this.name = name;
-		this.date = date;
-		this.reservationTime = reservationTime;
-	}
+    public Theme getTheme() {
+        return theme;
+    }
 
-	public ReservationResponse convert() {
-		return new ReservationResponse(id, name, date, reservationTime.getStartAt());
-	}
+    public boolean isBeforeThenNow() {
+        return LocalDateTime.of(date, reservationTime.getStartAt()).isBefore(LocalDateTime.now());
+    }
```
휘연님은 클래스마다 메서드를 어떻게 배치하시나요?
자신만의 컨벤션을 만들어봐도 좋을 것 같아요.
예를 들어 저는 getter 는 그리 중요한 로직이라고 생각하진 않아서 클래스의 밑 부분에 두고,
`isBeforeThenNow` 를 위에 두는 편이에요 (참고로 then -> than)
도메인은 애플리케이션에서 제일 중요한 도메인 로직을 담고 있으니, 잘 보이게 위로 올려보는건 어떨까요?

hhhhhwi replied:
생성자 코드 작성 후 바로 getter 를 작성하다보니 습관이 getter 메소드가 위로 가게끔 코드를 작성하고 있었네요 😓 
ㅎㅎ.. 이번 미션에서 유독 오탈자가 많네요 💦 
수정해보았습니다!

pkeugine commented on src/main/java/roomescape/reservation/dto/ReservationRequest.java:
```diff
@@ -0,0 +1,42 @@
+
+public class ReservationRequest {
+
+    @NotBlank(message = "이름")
```
이왕 메시지를 적는다면, 조금 더 구체적으로 적어보는건 어떨까요?

pkeugine commented on src/main/java/roomescape/reservation/service/ReservationRepository.java:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    public List<Reservation> find() {
+        String sql = new StringBuilder()
+            .append("SELECT ")
+            .append("r.id as reservation_id, ")
+            .append("r.name as reservation_name, ")
+            .append("r.date as reservation_date, ")
+            .append("t.id as time_id, ")
+            .append("t.start_at as time_start_at, ")
+            .append("th.id as theme_id, ")
+            .append("th.name as theme_name, ")
+            .append("th.description as theme_description, ")
+            .append("th.thumbnail as theme_thumbnail ")
+            .append("FROM reservation as r ")
+            .append("inner join reservation_time as t ")
+            .append("on r.time_id = t.id ")
+            .append("inner join theme as th ")
+            .append("on r.theme_id = th.id").toString();
```
string builder 를 사용해주셨네요!
이전 step 에서 제가 남긴 리뷰는 [multi-line string (또는 text block)](https://blogs.oracle.com/javamagazine/post/text-blocks-come-to-java) 을 사용해보자는 의미였습니다!

hhhhhwi replied:
헙 😅 
제가 잘못 이해했었군요! ㅎㅎ 반영해보았습니다!

pkeugine commented on src/main/java/roomescape/reservation/service/ReservationService.java:
```diff
@@ -0,0 +1,69 @@
+    private final ReservationTimeRepository reservationTimeRepository;
+    private final ThemeRepository themeRepository;
+
+    public ReservationService(ReservationRepository reservationRepository,
+        ReservationTimeRepository reservationTimeRepository, ThemeRepository themeRepository) {
```
```suggestion
    public ReservationService(ReservationRepository reservationRepository, 
                              ReservationTimeRepository reservationTimeRepository,
                              ThemeRepository themeRepository) {
```

줄이 길어졌을 때 파라미터는 이렇게 관리해봐도 괜찮을 것 같아요.

pkeugine commented on src/main/java/roomescape/reservation/service/ReservationService.java:
```diff
@@ -0,0 +1,69 @@
+            .collect(Collectors.toList());
+    }
+
+    public ReservationResponse findReservation(Long id) {
+        return new ReservationResponse(reservationRepository.findByKey(id));
+    }
```
이 메서드는 사용하지 않고 있는데, 삭제해도 괜찮다고 생각해요. 필요할 때 만들어주죠 :)

pkeugine commented on src/main/java/roomescape/reservation/service/ReservationService.java:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    public ReservationResponse saveReservation(ReservationRequest request) {
+        ReservationTime reservationTime = Optional.ofNullable(
+                reservationTimeRepository.findById(request.getTimeId()))
+            .orElseThrow(ReservationTimeNotExistsException::new);
+        Theme theme = Optional.ofNullable(themeRepository.findById(request.getThemeId()))
+            .orElseThrow(ThemeNotExistsException::new);
```
`Optional.ofNullable` 을 사용하시는 이유가 궁금해요!

hhhhhwi replied:
위 코드에서 `Optional.ofNullable`을 사용한 이유는 각 객체의 null 처리를 Exception 호출로 하기 위함이였습니다!

또한 `Optional`을 사용할 경우 반환 값이 null 이 될 수도 있음을 표현할 수 있으며 
다양한 메서드를 통해 null 처리를 보다 간단하게 구현할 수 있다는 점에서
선호하는 편입니다!

pkeugine reviewed (changes requested):
안녕하세요 휘연님, 잘 진행해주시고 계시네요!
부지런하게 작업하시는 모습 대단합니다... 👍 👍 
답변과 추가 코멘트 몇가지 남겨보았어요.
한 번 확인해보시면 될 것 같습니다 :)

hhhhhwi committed:
refactor: text blocks를 활용하여 multiline strings으로 긴 select 쿼리 정의

hhhhhwi committed:
refactor: 메소드명 오타 수정

hhhhhwi committed:
refactor: 에러메시지 수정 및 ReferenceException 상속 받는 CustomException 생성
- 에러메시지 코드 관리 Enum -> Class로 변경. 에러메시지를 상수로 정의하도록 수정
- 참조 오류 내용을 가진 ReferenceException을 상속하여 ReservationTimeReferenceException, ThemeReferenceException 생성

hhhhhwi commented on src/main/java/roomescape/theme/dto/ThemeRequest.java:
```diff
@@ -1,16 +1,17 @@
 public class ThemeRequest {
 
-    @NotBlank(message = "테마 이름")
+    @NotBlank(message = "테마 이름" + RoomescapeErrorMessage.BLANK_INPUT_VALUE_EXCEPTION)
```
> 이왕 메시지를 적는다면, 조금 더 구체적으로 적어보는건 어떨까요?

같은 Validation 어노테이션에 공통되지 않게 메시지를 정의를 하다보니, 
아.. 이게 유진님이 말씀하신 관리 포인트가 모여지지 않는다는 거구나.. 라는 생각이 들었습니다 ㅋㅋㅋㅋ 😅 

이제라도 각 Exception 마다 정의를 해서 관리하도록 할까..? 싶었지만 
CustomException 이외도 Handler를 통해 에러메시지를 재정의 하고 있어,  
고민 끝에 에러메시지를 Enum이 아닌 상수로 관리하도록 수정해보았습니다..!

pkeugine replied:
좋네요!
맞아요 예외 관련 정보를 한 곳에서 관리하는건 이상적이지만, 막상 유지보수하다보면 그렇게 한 곳에 모으기 쉽지가 않더라구요.
적절하게 바꿔주셨네요!! 👍

hhhhhwi committed:
refactor: ExceptionHandler 반환값 수정

pkeugine reviewed (approved):
안녕하세요 휘연님!
미션 정말 잘 진행해주셨네요.
이번 미션은 여기서 마무리하고, 다음 단계로 넘어가서 대화를 이어가보도록 하죠.
주말 잘 보내시고, 다음 미션도 파이팅입니다 :)
